### PR TITLE
Continuous Move/Rotate and CollisionListener

### PR DESCRIPTION
**Pull request by AlvaroHG**
Refactors collision checking with the arm to a new component CollisionListener that can cascade collision events up to parent CollisionListeners, to achieve a uniform way of rotating both the arm and the agent and paving the way to allow self intersection.

* Adds CollisionListener
* Refactors collision checking code
* Adds ContinuousMovement class for generic utility continuous Transform update functions
* Removes old code in arm, one single function for moving
* Adds MoveContinuous/MoveRotate
* Re-implements RotateMidLevelHand using this and latest collision checking

### CODE REVIEW

**Review by winthos — APPROVED** [review]

File: unity/Assets/Scripts/CollisionListener.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        public GameObject gameObject;
+
+        public SimObjPhysics simObjPhysics;
+
+         //indicates if gameObject a simObject
+        public bool isSimObj {
+            get { return simObjPhysics != null; }
+        }
+    }
+    private StaticCollision staticCollided;
```
Comment (ekolve) [inline unity/Assets/Scripts/CollisionListener.cs:24]: I think we can get rid of this.  It doesn't appear to be used.

File: unity/Assets/Scripts/CollisionListener.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                if(sop.PrimaryProperty == SimObjPrimaryProperty.Static)
+                {
+
+                    if(!col.isTrigger)
+                    {
+                        // #if UNITY_EDITOR
+                        // Debug.Log("Collided with static sim obj " + sop.name);
+                        // #endif
+                        sc = new StaticCollision();
+                        sc.simObjPhysics = sop;
```
Comment (ekolve) [inline unity/Assets/Scripts/CollisionListener.cs:86]: In the previous code only a simObj or gameObject was stored on the StaticCollision.  I don't think it matters too much, but it forces us to only do the null check where we check for a simobj first and then a gameobject in a certain order.  We only do that during debugging, so I don't think it changes the actual logic.

File: unity/Assets/Scripts/ContinuousMovement.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+using UnityEngine;
+using UnityStandardAssets.Characters.FirstPerson;
+using System;
+
+
+namespace UnityStandardAssets.Characters.FirstPerson {
+
+    public class ContinuousMovement {
+
+    public static void unrollSimulatePhysics(IEnumerator enumerator, float fixedDeltaTime) {
```
Comment (ekolve) [inline unity/Assets/Scripts/ContinuousMovement.cs:12]: Nice! This is what I was referring to when using the IEnumerable to unify the code.

**Review by ekolve — COMMENTED** [review]
I few comments, but otherwise looks good.

File: unity/Assets/Scripts/CollisionListener.cs (hunk 1 above)
Comment (winthos) [reply unity/Assets/Scripts/CollisionListener.cs:24]: agreed, this was from the previous version of the logic so should no longer be needed

Comment (AlvaroHG) [reply unity/Assets/Scripts/CollisionListener.cs:24]: Done, sneaked past me

File: unity/Assets/Scripts/CollisionListener.cs (hunk 2 above)
Comment (AlvaroHG) [reply unity/Assets/Scripts/CollisionListener.cs:86]: I wanted to simplify the check to not have an if that looked like `if (sc.simObjPhysics != null && sc.gameObject == null) { // then it is a simobject` which we had, so basically now if it is registered as a "StaticCollision" the gameObject will always be non-Null and there is a `isSimObject` field for quick checking if we collided against a SimObject.

The way to check if it is/is not a simObject you only call `sc.isSimObject`, and you avoid the null checking for the game object so it is the same one check, I just think is cleaner to have less nulls around and simplifies the above case. 
Does this make sense?

File: unity/Assets/Scripts/ContinuousMovement.cs (hunk 3 above)
Comment (AlvaroHG) [reply unity/Assets/Scripts/ContinuousMovement.cs:12]: Yeah good idea, thanks for mentioning it

**Pull request merged**